storage/src/main/java/org/apache/kafka/server/log/internals/ProducerAppendInfo.java

```diff
@@ -0,0 +1,238 @@
+                currentLastSeq = currentEntry.lastSeq();
+            else
+                currentLastSeq = RecordBatch.NO_SEQUENCE;
+
+            // If there is no current producer epoch (possibly because all producer records have been deleted due to
+            // retention or the DeleteRecords API) accept writes with any sequence number
+            if (!(currentEntry.producerEpoch() == RecordBatch.NO_PRODUCER_EPOCH || inSequence(currentLastSeq, appendFirstSeq))) {
+                throw new OutOfOrderSequenceException("Out of order sequence number for producer " + producerId + " at " +
+                        "offset " + offset + " in partition " + topicPartition + ": " + appendFirstSeq +
+                        " (incoming seq. number), " + currentLastSeq + " (current end sequence number)");
+            }
+        }
+    }
+
+    private boolean inSequence(int lastSeq, int nextSeq) {
+        return nextSeq == lastSeq + 1L || (nextSeq == 0 && lastSeq == Integer.MAX_VALUE);
+    }
+
+    public Optional<CompletedTxn> append(RecordBatch batch, Optional<LogOffsetMetadata> firstOffsetMetadataOpt) {
+        if (batch.isControlBatch()) {
+            Iterator<Record> recordIterator = batch.iterator();
+            if (recordIterator.hasNext()) {
+                Record record = recordIterator.next();
+                EndTransactionMarker endTxnMarker = EndTransactionMarker.deserialize(record);
+                return appendEndTxnMarker(endTxnMarker, batch.producerEpoch(), batch.baseOffset(), record.timestamp());
+            } else {
+                // An empty control batch means the entire transaction has been cleaned from the log, so no need to append
+                return Optional.empty();
+            }
+        } else {
+            LogOffsetMetadata firstOffsetMetadata = firstOffsetMetadataOpt.orElse(new LogOffsetMetadata(batch.baseOffset()));
+            appendDataBatch(batch.producerEpoch(), batch.baseSequence(), batch.lastSequence(), batch.maxTimestamp(),
+                    firstOffsetMetadata, batch.lastOffset(), batch.isTransactional());
+            return Optional.empty();
+        }
+    }
+
+    public void appendDataBatch(short epoch,
+                                int firstSeq,
+                                int lastSeq,
+                                long lastTimestamp,
+                                LogOffsetMetadata firstOffsetMetadata,
+                                long lastOffset,
+                                boolean isTransactional) {
+        long firstOffset = firstOffsetMetadata.messageOffset;
+        maybeValidateDataBatch(epoch, firstSeq, firstOffset);
+        updatedEntry.addBatch(epoch, lastSeq, lastOffset, (int) (lastOffset - firstOffset), lastTimestamp);
+
+        OptionalLong currentTxnFirstOffset = updatedEntry.currentTxnFirstOffset;
+        if (currentTxnFirstOffset.isPresent()) {
+            if (!isTransactional)
```

Comment:
Should this be moved into the previous `if` block?

Reply:
I preferred better readability here even though there is nesting. It looks little convoluted if we merge these nestings.

Reply:
How so? There are only two conditions in each block if we merge them?

Reply:
The current flow is if the currentTxnFirstOffset  optional is not present, it will always go to else block. But when we merge, even if optional is present and isTransaction true, it goes to else block, which is fine for now as it expects isTransactional in else block. I preferred the existing code for readability but I am fine with either ways. Let me change with what you suggested.  

```
if (currentTxnFirstOffset.isPresent()) {
    if (!isTransactional)
        // Received a non-transactional message while a transaction is active
        throw new InvalidTxnStateException("Expected transactional write from producer " + producerId + " at " +
                "offset " + firstOffsetMetadata + " in partition " + topicPartition);
} else {
    if (isTransactional) {
        // Began a new transaction
        updatedEntry.currentTxnFirstOffset = OptionalLong.of(firstOffset);
        transactions.add(new TxnMetadata(producerId, firstOffsetMetadata));
    }
}
```

```
if (currentTxnFirstOffset.isPresent() && !isTransactional) {
    // Received a non-transactional message while a transaction is active
    throw new InvalidTxnStateException("Expected transactional write from producer " + producerId + " at " +
            "offset " + firstOffsetMetadata + " in partition " + topicPartition);
} else if (isTransactional) {
    // Began a new transaction
    updatedEntry.currentTxnFirstOffset = OptionalLong.of(firstOffset);
    transactions.add(new TxnMetadata(producerId, firstOffsetMetadata));
}

```

Reply:
The current flow is if `currentTxnFirstOffset` is present, it should never go to the other else block. But when we merge the conditions, it will go to the else block incase of `currentTxnFirstOffset` is present and `isTransaction` true. This is a beaviour change which we do not want to have. So, it is better to keep the existing code as it is.

Reply:
@satishd Check the old code, it is equivalent to what I proposed here.

Reply:
Ah, I see what you mean now. The old code did:

```java
if (currentTxnFirstOffset.isPresent() && !isTransactional)
else if (!currentTxnFirstOffset.isPresent() && isTransactional)
```

I find that simpler to understand. There are two cases we care about and the rest we do not. The updated code as it stands hides that structure across two levels of nesting.

Reply:
Sure, that looks reasonable to me.